azure/src/main/scala/quasar/blobstore/azure/Azure.scala

```diff
@@ -16,93 +16,93 @@
-
-      case Some(AzureCredentials.ActiveDirectory(clientId, tenantId, clientSecret)) =>
-        rx.publisherToStream[F, AccessToken](
-          (new ClientSecretCredentialBuilder)
-            .clientId(clientId.value)
-            .tenantId(tenantId.value)
-            .clientSecret(clientSecret.value)
-            .tokenRefreshOffset(Duration.ofMinutes(60))
-            .build()
-            .getToken((new TokenRequestContext)
-              .setScopes(List("https://storage.azure.com/.default").asJava)))
-          .compile
-          .lastOrError
-          .map(tk => Expires(new TokenCredentials(tk.getToken), tk.getExpiresAt))
-    }
-
-  def mkContainerUrl[F[_]](cfg: Config)(implicit F: ConcurrentEffect[F]): F[Expires[ContainerURL]] =
-    F.catchNonFatal(new URL(cfg.storageUrl.value)) flatMap { url =>
-      // Azure SDK changed its logging behavior in 10.3.0 (compared to 10.2.0).
-      // In 10.2.0 the standard `new PipelineOptions` could be used, but in 10.3.0
-      // this results in messages being logged to console in eg quasar's repl
-      // (despite having no console logger configured).
-      // Still not having a full understanding of how this logging is supposed to work
-      // exactly (e.g. `withLogger`) but adding `LoggingOptions` with
-      // `disableDefaultLogging = true` seems to be fixing it.
-      // As an aside, in case we need to look at this in the future,
-      // Azure SDK has 2 loggers for some reason:
-      // - `com.microsoft.azure.storage.blob.LoggingFactory`
-      // - `Azure Storage Java SDK`
-
-      Functor[F].compose[Expires].map(mkCredentials(cfg.credentials)) { creds =>
-        val pipeline =
-          StorageURL.createPipeline(
-            creds, new PipelineOptions().withLoggingOptions(new LoggingOptions(3000, true)))
-
-        new ServiceURL(url, pipeline).createContainerURL(cfg.containerName.value)
+  def getAccessToken[F[_]: Async: ContextShift](ad: ActiveDirectory): F[Expires[AccessToken]] =
+    Sync[F].delay {
+      (new ClientSecretCredentialBuilder)
+        .clientId(ad.clientId.value)
+        .tenantId(ad.tenantId.value)
+        .clientSecret(ad.clientSecret.value)
+        .tokenRefreshOffset(Duration.ofMinutes(60))
+        .build()
+        .getToken((new TokenRequestContext)
+          .setScopes(List("https://storage.azure.com/.default").asJava))
+    } flatMap reactive.monoToAsync[F, AccessToken] map (t => Expires(t, t.getExpiresAt()))
+
+  def setCredentials[F[_]: ConcurrentEffect: ContextShift](
+      cred: AzureCredentials,
+      builder: BlobContainerClientBuilder)
+      : F[Expires[BlobContainerClientBuilder]] =
+    ConcurrentEffect[F].delay {
+      cred match {
+        case AzureCredentials.SharedKey(accountName, accountKey) =>
+          Expires.never(
+            builder
+              .credential(new StorageSharedKeyCredential(accountName.value, accountKey.value))).pure[F]
+        case ad@AzureCredentials.ActiveDirectory(_, _, _) =>
```

Comment:
Trying not to style nitpick too much, but… it's very important to *not* do pattern matching this way. `case ad @ AzureCredentials.ActiveDirectory(_, _, _) =>` is much, much safer because no-space use of `@` can trip the parser's annotation precedence.

Reply:
Ah ok, thanks. I didn't know that.